src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,271 @@
+  private JdbcConnection writerConnection;
+  private JdbcConnection readerConnection;
+
+  public ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this(currentConnectionProvider, new BasicConnectionProvider(), propertySet, nextPlugin, logger);
+  }
+
+  ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionProvider connectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.connectionProvider = connectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+  }
+
+  @Override
+  public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc, Object[] args) throws Exception {
+    if(METHOD_SET_READ_ONLY.equals(methodName)) {
+      switchConnectionIfRequired((boolean) args[0]);
+    }
+    return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+  }
+
+  void switchConnectionIfRequired(boolean readOnly) throws SQLException {
+    JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (readOnly) {
+      if (!isReaderConnection() || currentConnection.isClosed()) {
+        try {
+          switchToReaderConnection(currentConnection);
+        } catch (SQLException e) {
+          if(currentConnection == null || currentConnection.isClosed()) {
+            throw new SQLException(
+                Messages.getString("ReadWriteSplittingPlugin.1"),
+                MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e);
+          }
+
+          // stick with the current connection
+          if (this.readerConnection != currentConnection) {
+            if (this.readerConnection != null) {
+              this.readerConnection.close();
+            }
+            this.readerConnection = currentConnection;
+          }
+        }
+      }
+    } else {
+      if (!isWriterConnection() || currentConnection.isClosed()) {
+        try {
+          switchToWriterConnection(currentConnection);
+        } catch (SQLException e) {
+          throw new SQLException(
+              Messages.getString("ReadWriteSplittingPlugin.2"),
+              MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e);
+        }
+      }
+    }
+    this.readOnly = readOnly;
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    this.hosts.addAll(connectionUrl.getHostsList());
+    this.nextPlugin.openInitialConnection(connectionUrl);
+    this.writerConnection = this.currentConnectionProvider.getCurrentConnection();
+  }
+
+  @Override
+  public void releaseResources() {
+    this.nextPlugin.releaseResources();
```

Comment:
good point... shouldn't we close both connections in this case or no? I don't see releaseResources being called in the ConnectionPluginManager so I'm not sure under which scenarios this would be called

Reply:
The question is more about ownership of an additional connection. Current connection is owned by connection proxy while additional connections, like a reader connection in this plugin, is owned by plugin. Thus, it's a responsibility of the plugin to close and dispose all additional resources.

Reply:
Right, that makes sense. At the time releaseResources is called the plugin could have set the current connection to be either the reader or the writer, so I'll have it check and close any connections that aren't the current connection